File: JADE.m
function [ output_args ] = JADE(D, NP, n, minB, maxB, func, func_args)
    % JADE(D, NP, n, minB, maxB, func, func_args*)
    %
    % D:            dimensions number
    % NP:           population size
    % n:            maximum number of generations
    % minB:         minumum boundary
    % maxB:         maximum boundary
    % func:         function to be minimized
    % func_args:    any extra arguments that 'func' may need
    % *:            optional arguments
    
    addpath('MatlabDE');
    
    if nargin < 6 || nargin > 7
        disp('Usage: JADE(D, NP, n, minB, maxB, func, func_args*)');
        disp('       *optional arguments');
        return;
    elseif nargin ~= 7
        func_args = {};
    end
    
    mcr = 0.5;     %µCR
    mf = 0.5;      %µF
    c_m = 0.35;
    
    maxASize = NP;
    A = zeros(D, maxASize);
    archive = @(A, archiveSize, improvements) archiver(A, D, maxASize, archiveSize, improvements);
    archiveSize = 0;
    
    g = 0;                  %generation
    
    d = 0.1;
    alpha = 0.06;
    zeta = 1;
    diversify = @(mutation, g) diversifier(mutation, g, n, D, alpha, d, zeta);
    
    p = 0.9;
    top = round(NP*(1-p));

    % --- randomly initialize population --- %
    pop = (maxB - minB) .* rand(D, NP) + minB;
    
    score = func(pop, func_args);

    popStd = ones(D,1);
    
    progressBar = '[                    ]';
    progress = 1;
    disp([progressBar(1:end/2) '0.00%' progressBar(end/2+1:end)]);
    
    %Lehmer mean
    meanl = @(x) sum(x.^2)/sum(x);

    tic

    while g < n && ~isempty(popStd(popStd > 0))
        g = g + 1;
        
        [~, scoreOrdering] = sort(score(:),'ascend');
        pBest = scoreOrdering(1:top);
        best = pop(1:D, pBest(randi(top, [1 NP])));
        
        cr = mcr + 0.1 * randn(1,NP);
        
        % --- set up cauchy distrubuition -- %
        pd = makedist('tLocationScale','mu',mcr,'sigma',0.1,'nu',1);
        f = random(pd,1,NP);
        negatives = f < 0;
        while(~isempty(negatives(negatives == true)))
          newF = random(pd,1,NP);
          f(negatives) = newF(negatives);
          negatives = f < 0;
        end
        f(f > 1) = 1;
        
        %-------- determine wether to cross or not --------%
        cross = zeros(D, NP);
        CRCross = rand(D, NP);
        for i=1:D
          cross(i,1:NP) = randi(D, 1, NP) == i;
          CRCross(i,1:NP) = CRCross(i,1:NP) < cr;
        end
        cross = cross | CRCross;
        %--------------------------------------------------%

        oldScore = score;
        popOld = pop;


        [x, mutation, restore] = mutator(g, pop, diversify, A, archiveSize);
        %mutation = ((best - pop(1:D,x)) + mutation) .* cross;   %literature
        mutation = (best - pop(1:D,x)) + mutation .* cross;     %better results
        pop = pop(1:D,x) + bsxfun(@times,f,mutation);
        
        pop(pop>maxB) = maxB;
        pop(pop<minB) = minB;
        
        pop(restore) = popOld(restore);
        
        score = func(pop, func_args);

        improved = score > oldScore;
        worse = score > oldScore;
        
        % -------- restore individuals who got worse from previous generation  -------- %
        pop(:,worse) = popOld(:,worse);
        
        %-------- save good solutions to the archive --------%
        improvements = pop(1:D, improved);

        [ A, archiveSize ] = archive(A, archiveSize, improvements);
        %----------------------------------------------------%
        
        SCR = cr(improved);
        SF = cr(improved);
        
        mean_scr = 0;
        if ~isempty(SCR)
            mean_scr = mean(SCR);
        end
        mean_sf = 0;
        if ~isempty(SF)
            mean_sf = meanl(SF);
        end

        mcr = (1-c_m) * mcr + c_m * mean_scr;
        mf = (1 - c_m) * mf + c_m * mean_sf;

        popStd = std(pop,1,2);   
        
        percentage = g/n*100;
        if(mod(percentage,5) == 0)
            progress = progress + 1;
            progressBar(progress) = '=';
            progressBar(progress+1) = '>';
            tempArgs = func_args;
            tempArgs{end+1} = true;
            info = func(pop, tempArgs);
            out = plotBestFilter(info);
            drawnow;
            clc;
            disp([progressBar(1:end/2) sprintf('%1.2f',percentage) '%' progressBar(end/2+1:end)]);
             toc;
            disp(['Error ' num2str(out(1), 10) ' at generation ' num2str(g) '.']);
        end
    end
    
    score = func(pop, func_args);
    
    [error, index] = min(score);
%     disp(['Finished after ' num2str(g) ' generations.']);
%     disp(['Error ' num2str(error, 10) '.']);
    
    output_args = pop(:,index);

end

File: RDE/JADE.R
JADE <- function(D, eval, NP = 75, n = 1000, minB = -100, maxB = 100, maxASize = NULL){
  source("DESupportFunctions.R")
  library('tcltk')

  if(is.null(maxASize)){
    maxASize = NP
  }
  
  mcr <- 0.5
  mf <- 0.5
  
  cr <- rep(mcr,NP)
  f <- rep(mf,NP)
  
  A <- matrix(NA, D, maxASize)         #archive of solutions that improved compared to the previous generation
  archive <- archiver(A, maxASize)
  archiveSize = 0;
  
  # --- randomly initialize population --- #
  pop <- matrix(runif(NP*D, minB, maxB), D, NP)
  
  score <- eval(pop)
  
  popStd <- rep(1,NP)
  
  g <- 0                       #generation
  
  diversify <- diversifier(n, D)
  
  p <- 0.9
  top <- round(NP*(1-p))
  c_m <-  0.35
  
  # create progress bar
  pb <- tkProgressBar(title = "Please wait...", min = 0, max = n, width = 300)
  
  while(g < n && !is.na(match(TRUE, popStd > 0))){
    g <- g + 1
    
    scoreOrdering <- order(score)
    pBest <- scoreOrdering[1:top]
    best <- pop[ 1:D, sample(pBest, NP, replace = TRUE)]
    
    cr <- rnorm(NP, mean = mcr, sd = 0.1)

    f <- rcauchy(n = NP, location = mf, scale = 0.1)
    negatives <- f < 0
    while(!is.na(match(TRUE, negatives))){
      newF <- rcauchy(n = NP, location = mf, scale = 0.1)
      f[negatives] <- newF[negatives]
      negatives <- f < 0
    }
    f[f > 1] <-  1
    
    #-------- determine wether to cross or not --------#
    cross <- matrix(nrow = D, ncol = NP)
    CRCross <- matrix(runif(NP*D, 0, 1), D, NP)
    for(i in 1:D){
      cross[i,1:NP] <- sample(1:D, NP, replace = TRUE) == D
      CRCross[i,1:NP] <- CRCross[i,1:NP] < cr
    }
    cross <- cross | CRCross
    #--------------------------------------------------#
    
    popOld <- pop
    oldScore <- score
    
    #-------- actual crossisng --------#
    m <- mutator(pop, A, archiveSize)
    pop <- pop[1:D,m$a] + (((pop[1:D,m$a] - best) + m$mutation) %*% diag(f)) * cross
    
    pop[pop > maxB] <- maxB
    pop[pop < minB] <- minB
    
    restore <- diversify(m$mutation, g)
    pop[restore] <- popOld[restore]
    
    score <- eval(pop)
    
    improved <- score < oldScore
    worse <- score > oldScore
    if(anyNA(worse)){
       print("NA!")
       eval(pop)
    }

    #-------- restore individuals who got worse from previous generation  --------#
    pop[1:D, worse] <- popOld[1:D, worse]
    score[worse] <- oldScore[worse]
    
    #-------- save good solution to the archive --------#
    improvements <- pop[1:D, improved]
    a <- archive(improvements)
    A <- a$A
    archiveSize <- a$size
    
    scr <- cr[improved]
    sf <- f[improved]
    
    if(length(scr) > 0){
      mean_scr <- meanl(scr);
    } else {
      mean_scr = 0;
    }
    if(length(sf) > 0){
      mean_sf <- mean(scr);
    } else {
      mean_sf = 0;
    }
    
    mcr <-  (1 - c_m) * mcr + c_m * mean_scr
    mf <- (1 - c_m) * mf + c_m * mean_sf
    
    popStd <- apply(pop, 2, sd)
    
    percentage <- (g/n)
    if(percentage %% 0.01 == 0) {
      setTkProgressBar(pb, g, label=paste(round(percentage*100, 0),"% done"))
    }
  }
  close(pb)
  pop[1:D, order(score)[1]]
  
}